### PR TITLE
Improve data collection of structured deployment tracing

### PR DESCRIPTION
**Pull request by pdudits**
Use consistent container names, document what I think the individual actions represent and compute nesting levels of individual spans. Current trace output looks like this: (without the headers)

```
Start			Finish			Nest  Container	Context				Action  Component	Duration
16:05:03.268000000	16:08:52.827000000	1		APPLICATION dataverse.war	null		229559
16:05:03.268000000	16:05:03.269000000	2		APPLICATION dataverse.war	OPENING_ARCHIVE		1
16:05:03.269000000	16:05:03.269000000	2		APPLICATION dataverse.war	VALIDATE_TARGET	command	0
16:05:03.269000000	16:05:03.270000000	2		APPLICATION dataverse.war	OPENING_ARCHIVE	ArchiveHandler	1
16:05:03.270000000	16:05:03.270000000	2		APPLICATION dataverse.war	CREATE_DEPLOYMENT_CONTEXT	Initial	0
16:05:03.270000000	16:05:03.270000000	2		APPLICATION dataverse.war	PROCESS_EVENTS	Initial_Context_Created	0
16:05:03.270000000	16:05:03.311000000	2		APPLICATION dataverse.war	DETERMINE_APP_NAME		41
16:05:03.270000000	16:05:03.270000000	3		APPLICATION dataverse.war	DETERMINE_APP_NAME	ApplicationInfoProvider	0
16:05:03.270000000	16:05:03.310000000	3		APPLICATION dataverse.war	DETERMINE_APP_NAME	DeploymentDescriptors	40
16:05:03.313000000	16:05:03.313000000	2		APPLICATION dataverse.war	VALIDATE_TARGET	registry	0
16:05:03.313000000	16:08:08.196000000	2		APPLICATION dataverse.war	CLEANUP	applications	184883
16:08:08.196000000	16:08:08.196000000	2		APPLICATION dataverse.war	SWITCH_VERSIONS		0
16:08:08.197000000	16:08:10.267000000	2		APPLICATION dataverse.war	CREATE_DEPLOYMENT_CONTEXT	Full	2070
16:08:10.267000000	16:08:10.272000000	2		APPLICATION dataverse.war	CLEANUP	generated	5
16:08:10.272000000	16:08:10.272000000	2		APPLICATION dataverse.war	PREPARE	ServerConfig	0
16:08:10.272000000	16:08:10.272000000	2		APPLICATION dataverse.war	PROCESS_EVENTS	Deployment_Start	0
16:08:10.272000000	16:08:42.325000000	2		APPLICATION dataverse.war	PREPARE		32053
16:08:10.272000000	16:08:10.273000000	3		APPLICATION dataverse.war	PREPARE	ArchiveMetadata	1
16:08:10.273000000	16:08:10.273000000	3		APPLICATION dataverse.war	PREPARE	ArchiveHandler	0
16:08:10.273000000	16:08:16.198000000	3		APPLICATION dataverse.war	CLASS_SCANNING		5925
16:08:10.273000000	16:08:10.834000000	4		APPLICATION dataverse.war	CLASS_SCANNING	dataverse	561
16:08:16.198000000	16:08:16.249000000	3		APPLICATION dataverse.war	PREPARE	Sniffer	51
16:08:16.249000000	16:08:16.249000000	3		APPLICATION dataverse.war	PREPARE	ClassLoaderHierarchy	0
16:08:16.249000000	16:08:16.540000000	3		APPLICATION dataverse.war	PREPARE	ClassLoader	291
16:08:16.540000000	16:08:19.085000000	3		APPLICATION dataverse.war	PREPARE	Container	2545
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	jpa	APPLICATION dataverse.war > CONTAINER jpa	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	jpa	APPLICATION dataverse.war > CONTAINER jpa	PREPARE	Deployer	0
16:08:16.540000000	16:08:16.540000000	4	web	APPLICATION dataverse.war > CONTAINER web	PREPARE		0
16:08:16.540000000	16:08:16.540000000	4	web	APPLICATION dataverse.war > CONTAINER web	PREPARE	Deployer	0
16:08:16.540000000	16:08:19.085000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		2545
16:08:16.540000000	16:08:19.085000000	5		APPLICATION dataverse.war	LOAD	DeploymentDescriptor	2545
16:08:19.085000000	16:08:19.085000000	3	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE		0
16:08:19.085000000	16:08:19.085000000	3	security	APPLICATION dataverse.war > CONTAINER security	PREPARE		0
16:08:19.085000000	16:08:19.085000000	3	jpa	APPLICATION dataverse.war > CONTAINER jpa	PREPARE		0
16:08:19.085000000	16:08:19.085000000	3	web	APPLICATION dataverse.war > CONTAINER web	PREPARE		0
16:08:19.085000000	16:08:19.085000000	3	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE	MetaData	0
16:08:19.085000000	16:08:19.085000000	3	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE	MetaData	0
16:08:19.085000000	16:08:19.085000000	3	security	APPLICATION dataverse.war > CONTAINER security	PREPARE	MetaData	0
16:08:19.085000000	16:08:19.085000000	3	jpa	APPLICATION dataverse.war > CONTAINER jpa	PREPARE	MetaData	0
16:08:19.085000000	16:08:19.085000000	3	web	APPLICATION dataverse.war > CONTAINER web	PREPARE	MetaData	0
16:08:19.085000000	16:08:19.499000000	3	web	APPLICATION dataverse.war > CONTAINER web	CREATE_CLASSLOADER		414
16:08:19.499000000	16:08:23.158000000	3	web	APPLICATION dataverse.war > CONTAINER web	PREPARE	Module	3659
16:08:19.499000000	16:08:19.499000000	4	cdi	APPLICATION dataverse.war > CONTAINER cdi	PREPARE		0
16:08:19.499000000	16:08:19.502000000	4	ejb	APPLICATION dataverse.war > CONTAINER ejb	PREPARE		3
16:08:19.502000000	16:08:19.503000000	4	security	APPLICATION dataverse.war > CONTAINER security	PREPARE		1
16:08:19.503000000	16:08:23.156000000	4	jpa	APPLICATION dataverse.war > CONTAINER jpa	PREPARE		3653
16:08:23.156000000	16:08:23.157000000	4	web	APPLICATION dataverse.war > CONTAINER web	PREPARE		1
16:08:23.157000000	16:08:23.158000000	4	web	APPLICATION dataverse.war > CONTAINER web > MODULE dataverse	PROCESS_EVENTS	Module_Prepared	1
16:08:23.158000000	16:08:24.557000000	3	web	APPLICATION dataverse.war > CONTAINER web > MODULE dataverse	PROCESS_EVENTS	Application_Prepared	1399
16:08:24.557000000	16:08:42.325000000	3	web	APPLICATION dataverse.war > CONTAINER web > MODULE dataverse	LOAD		17768
16:08:24.557000000	16:08:39.095000000	4		APPLICATION dataverse.war	LOAD		14538
16:08:24.557000000	16:08:24.557000000	5		APPLICATION dataverse.war > MODULE dataverse	LOAD		0
16:08:24.557000000	16:08:39.087000000	5		APPLICATION dataverse.war > MODULE dataverse	LOAD		14530
16:08:24.557000000	16:08:33.374000000	6	cdi	APPLICATION dataverse.war > MODULE dataverse > CONTAINER cdi	LOAD		8817
16:08:33.374000000	16:08:39.087000000	6	ejb	APPLICATION dataverse.war > MODULE dataverse > CONTAINER ejb	LOAD		5713
16:08:39.087000000	16:08:39.087000000	5	security	APPLICATION dataverse.war > MODULE dataverse > CONTAINER security	LOAD		0
16:08:39.087000000	16:08:39.087000000	5	jpa	APPLICATION dataverse.war > MODULE dataverse > CONTAINER jpa	LOAD		0
16:08:39.087000000	16:08:39.087000000	5	web	APPLICATION dataverse.war > MODULE dataverse > CONTAINER web	LOAD		0
16:08:39.095000000	16:08:42.325000000	4		APPLICATION dataverse.war	PROCESS_EVENTS	Application_Loaded	3230
16:08:42.325000000	16:08:42.325000000	2		APPLICATION dataverse.war	INITIALIZE		0
16:08:42.325000000	16:08:52.762000000	2		APPLICATION dataverse.war	START		10437
16:08:42.325000000	16:08:52.827000000	3		APPLICATION dataverse.war	START		10502
16:08:42.325000000	16:08:42.325000000	4		APPLICATION dataverse.war > MODULE dataverse	START		0
16:08:42.325000000	16:08:42.325000000	4		APPLICATION dataverse.war > MODULE dataverse	PROCESS_EVENTS	Module_Running	0
16:08:42.325000000	16:08:52.827000000	4		APPLICATION dataverse.war > MODULE dataverse	START		10502
16:08:42.325000000	16:08:52.762000000	5		APPLICATION dataverse.war > MODULE dataverse	START		10437
16:08:42.325000000	16:08:42.325000000	6	cdi	APPLICATION dataverse.war > MODULE dataverse > CONTAINER cdi	START		0
16:08:42.325000000	16:08:42.355000000	6	ejb	APPLICATION dataverse.war > MODULE dataverse > CONTAINER ejb	START		30
16:08:42.355000000	16:08:42.355000000	6	security	APPLICATION dataverse.war > MODULE dataverse > CONTAINER security	START		0
16:08:42.355000000	16:08:42.355000000	6	jpa	APPLICATION dataverse.war > MODULE dataverse > CONTAINER jpa	START		0
16:08:42.355000000	16:08:52.762000000	6	web	APPLICATION dataverse.war > MODULE dataverse > CONTAINER web	START		10407
````

### CODE REVIEW

**Comment by arjantijms** [comment]
Jenkins test please

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java [hunk 1]
```diff
@@ -885,7 +928,7 @@ private EngineInfo startEngine(DeploymentContext context, Sniffer sniffer, Strin
         StructuredDeploymentTracing tracing = StructuredDeploymentTracing.load(context);
 
         // start all the containers associated with sniffers.
-        try (DeploymentSpan span = tracing.startSpan(TraceContext.Level.CONTAINER, containerName, DeploymentTracing.AppStage.PREPARE)) {
+        try (DeploymentSpan span = tracing.startSpan(TraceContext.Level.CONTAINER, sniffer.getModuleType(), DeploymentTracing.AppStage.PREPARE)) {
```
Comment (jGauravGupta) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java]: Shouldn't "Engine" component type be added here?

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java [hunk 2]
```diff
@@ -520,15 +560,17 @@ public void initialize(ApplicationInfo appInfo, Collection<? extends Sniffer> sn
         appRegistry.removeTransient(appInfo.getName());
         final ActionReport report = context.getActionReport();
         ProgressTracker tracker = context.getTransientAppMetaData(ExtendedDeploymentContext.TRACKER, ProgressTracker.class);
+        StructuredDeploymentTracing tracing = StructuredDeploymentTracing.load(context);
         // now were falling back into the mainstream loading/starting sequence, at this
         // time the containers are set up, all the modules have been prepared in their
         // associated engines and the application info is created and registered
         if (loadOnCurrentInstance(context)) {
-            try {
+            try (SpanSequence span = tracing.startSequence(DeploymentTracing.AppStage.INITIALIZE)){
```
Comment (jGauravGupta) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java]: Is START app stage part of INITIALIZE SpanSequence?

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/ApplicationInfo.java [hunk 3]
```diff
@@ -321,7 +321,7 @@ public void load(ExtendedDeploymentContext context, ProgressTracker tracker)
         span.close();
 
         if (events!=null) {
-            span = tracing.startSpan(TraceContext.Level.APPLICATION, name, DeploymentTracing.AppStage.LOAD_EVENTS);
+            span = tracing.startSpan(TraceContext.Level.APPLICATION, name, DeploymentTracing.AppStage.PROCESS_EVENTS, Deployment.APPLICATION_LOADED.type());
             events.send(new Event<ApplicationInfo>(Deployment.APPLICATION_LOADED, this), false);
```
Comment (jGauravGupta) [inline nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/ApplicationInfo.java]: Please also update the tracing snippet used in ApplicationInfo#start method:

````
DeploymentSpan innerSpan = tracing.startSpan(TraceContext.Level.APPLICATION, getName(), DeploymentTracing.AppStage.PROCESS_EVENTS, Deployment.APPLICATION_STARTED.type());
events.send(new Event<ApplicationInfo>(Deployment.APPLICATION_STARTED, this), false);
````

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/ModuleInfo.java [hunk 4]
```diff
@@ -287,7 +287,7 @@ public synchronized void start(
             }
             started=true;
             if (events!=null) {
-                DeploymentSpan innerSpan = tracing.startSpan(DeploymentTracing.AppStage.START_EVENTS, "Module");
+                DeploymentSpan innerSpan = tracing.startSpan(DeploymentTracing.AppStage.PROCESS_EVENTS, Deployment.MODULE_STARTED.type());
                 events.send(new Event<ModuleInfo>(Deployment.MODULE_STARTED, this), false);
```
Comment (jGauravGupta) [inline nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/ModuleInfo.java]: Shouldn't MODULE_LOADED event also be traced like MODULE_STARTED, APPLICATION_LOADED, and APPLICATION_STARTED.

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java [hunk 5]
```diff
@@ -441,7 +481,7 @@ public void actOn(Logger logger) {
                 return null;
             }
 
-            span.start(DeploymentTracing.AppStage.PREPARE, "Notifying");
+            span.start(DeploymentTracing.AppStage.PROCESS_EVENTS, Deployment.APPLICATION_PREPARED.type());
```
Comment (jGauravGupta) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java]: `events.send(new Event<DeploymentContext>(Deployment.APPLICATION_PREPARED, context), false)` invoked on line 517. Shouldn't tracer snippet move there?

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/deployment/analysis/DeploymentSpan.java [hunk 6]
```diff
@@ -115,9 +117,15 @@ public String toString() {
...
+        sb.append('\t')
+            .append(getNestLevel())
+            .append('\t');
         if (context != null) {
-            sb.append(context);
+            sb.append(context.getLevelName(TraceContext.Level.CONTAINER))
+                .append('\t')
+                .append(context);
+        } else {
+            sb.append('\t');
```
Comment (jbee) [inline nucleus/common/internal-api/src/main/java/org/glassfish/internal/deployment/analysis/DeploymentSpan.java]: Is tab really the right way to do this? I think the current output example shows that it mostly lines up but sometimes it is 1-2 tabs out of alignment. Wouldn't use of format with fixed with spacing be more accurate. Also for things like missing "columns" like the case here where context sometimes is empty/non existent?

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/deployment/analysis/TraceContext.java [hunk 7]
```diff
@@ -96,6 +96,16 @@ public String getName() {
         return name;
     }
 
+    public String getLevelName(Level level) {
+        if (level == this.level) {
+            return name;
+        } else if (parent != null) {
```
Comment (jbee) [inline nucleus/common/internal-api/src/main/java/org/glassfish/internal/deployment/analysis/TraceContext.java]: `else` cascade is not needed when using return. I know our formatting of always using curly brackets makes it longer if a sequence of individual `if` statements is used but I still think its more readable.

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/deployment/analysis/DeploymentSpan.java (hunk 6 above)
Comment (pdudits) [reply nucleus/common/internal-api/src/main/java/org/glassfish/internal/deployment/analysis/DeploymentSpan.java]: Tabs import well into spreadsheet, which is the currently used "frontend' for the feature. The output is too complex to just inspect visually in the log, so it's adapted for import by some tooling.

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java (hunk 1 above)
Comment (pdudits) [reply nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java]: I considered tracking engines separately, but it didn't help the clarity of the result. The engines are tightly coupled with their respective containers, so the distinction doesn't help. (See also this comment above, that mentions sniffers and containers).

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/ApplicationInfo.java (hunk 3 above)
Comment (pdudits) [reply nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/ApplicationInfo.java]: Will do.

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java (hunk 5 above)
Comment (pdudits) [reply nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java]: Good catch, thanks

**Comment by pdudits** [comment]
jenkins test please

**Pull request merged**